lib/libzfs/libzfs_pool.c

```diff
@@ -3854,6 +3854,10 @@ zpool_history_unpack(char *buf, uint64_t bytes_read, uint64_t *leftover,
 		if (ISP2(*numrecords + 1)) {
 			*records = realloc(*records,
 			    *numrecords * 2 * sizeof (nvlist_t *));
+			if (*records == NULL) {
```

Comment:
You need to reintroduce the the `tmp` pointer. Callers are responsible for freeing the nvlist array pointed at by `*records`. If `realloc` returned `NULL`, then free the nvlist we alloced and return `ENOMEM`. Otherwise, do `*records = tmp;`.

Also, `numrecords` above is being incremented before the failure. If you hit this case, that means `numrecords` is too large by 1. I'm can't think of a "clean" way to do that. Some options are to store the old value of `numrecords`, don't increment `numrecords` until we pass all failure cases, or increment then decrement on failure. 

@behlendorf thoughts?